---
title: "Vignette"
author: "Mike Sportiello"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Abstract

Efficient breeding is a scientific imperative: with a limited period of fertility in experimental organisms, users may only have limited attempts with organisms of desired genotypes [@RN204] [@RN203] since the time of fertility is often the limiting factor in experimental organisms. An inefficient cross can set the user back months in the case of a female pregnant with the non-optimal male's offspring once both gestational time and the necessary time it takes to wean pups off their mother's milk. Furthermore, as most research centers pay per diem costs, inefficiency is a large direct cost for many experimentalists, with individual cages often costing hundreds of dollars per year just in maintenance fees, not even accounting for the cost of extra rounds of genotyping for which one must account for genotyping costs (PCR primers, PCR mastermix, gels, dye, etc) or for people's time, which may take the form of decreased data production and the cost of of actually paying employees to do this work. Ethical and animal welfare concerns also arise as inefficient breeding results in the mass euthanasia of non-desired animals as well as ear/tail-clipping required to genotype which may cause animals pain and distress.

While the breeding schema to obtain a litter of full knockout mice from one Wildtype/Wildtype parent and one knockout/knockout parent is straightforward, this is not the case for when one has 4 alleles at 4 loci in 4 separate mice that need to be in the same mouse for a given planned experiment, for example. Furthermore, no current software exists to our knowledge that can quantitatively assist the user in creating their breeding schema. Indeed, if there are 2 alleles at each locus, when attempting to make a genetically marked, inducible, cre-lox model with T cell specificity as our lab was doing, prompting us to create this software, 81 possible combinations exist (3 (AA, Aa, and aa) to the 4th power). With 81 possible males and 81 possible females to mate, 6561 possible pairings exist (81 * 81). Use of this mousebreedeR software to optimize that breeding schema resulted in the publication of a thesis project and manuscript being prepped for submission. 

# Walkthrough
First we start off by loading all of the necessary libraries.
```{r message=FALSE, warning=FALSE}
library(mousebreedeR)
library(dqrng)
library(dplyr)
library(gtools)
library(ggplot2)
library(reshape2)
library(viridis)
library(ggpubr)
`%!in%` <- Negate(`%in%`)
```

The package takes a table of mouse genotypes and gives the user breeding strategy. The user supplies a data frame with a numeric number of copies of a certain allele. For example, for the allele A, the genotype AA would have 2 copies, Aa 1 copy, and aa 0 copies. The data frame supplied should use actual numbers, not phrases like "homozygous positive" or "AA," for example. 

In the below code chunk, we can build an example data frame. The last argument `head(df)` will show a properly formatted input to the package, but most people will find it easier to read in a csv.

YOU CAN REPLACE `df` WITH YOUR DATAFRAME OF INTEREST AND RUN THE BELOW CODE. MAKE SURE TO ALSO UPDATE THE `desiredvec` BELOW APPROPRIATELY.
```{r}
# make example df
gene1 <- c(0, 0, 0, 1, 0, 0, 0, 1)
gene2 <- c(0, 0, 1, 0, 0, 0, 1, 0)
gene3 <- c(0, 1, 0, 0, 0, 1, 0, 0)
gene4 <- c(1, 0, 0, 0, 1, 0, 0, 0)
gene5 <- c('M', 'M', 'M', 'M','F', 'F', 'F', 'F')
df <- data.frame(gene1, gene2, gene3, gene4, gene5)
colnames(df) <- c(paste0('gene', 2:dim(df)[2] - 1), 'sex')
head(df)
```

SAVE YOUR DATAFRAME OF AVAILABLE MICE TO THE VARIABLE df BELOW. This is how the program knows what mice you have as starting mice. DON'T FORGET TO UPDATE desiredvec BELOW (line 115). You can change the path in the below function to any csv file you want, but it should look like that format. Feel free to add rows and columns as you see fit. 
```{r}

# df<-read.csv('edit_this_CSV_with_breeder_mice.csv') 
```

```{r}
# rename df
exampleData<-df
```

As you can see above, the format of the data frame should be as follows: each column is an allele (A or B or C, for example), each row is a mouse. Each cell should have the number of that allele in that mouse. For a mouse that is AABbCcdd, the row should have 2 in the first column, 1 in the second column, 1 in the third column, and 0 in the fourth column. The last column should be the sex of the animal: M for male, and F for female.

In the next code chunk, we set the theme for the plots we make later on.
```{r}
# set ggplothemes
theme_set(
  theme(
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 8),
    strip.text = element_text(size = 8),
    axis.text.x = element_text(angle = 90),
    legend.position = 'none',
    # strip.text.x = element_text(size = 8,margin = margin(.1, 0, .1, 0, "cm")),
    legend.text = element_text(size = 7),
    legend.title = element_text(size = 7),
    # legend.position = 'bottom',panel.border=element_blank(),
    panel.background = element_blank(),
    panel.grid.major = element_line(color = 'light grey')
  )
)
```

First let's look at what happens when each mouse undergoes meiosis and creates gametes. Then, we can look at the possible gametes produced by the mice from the data frame you originally created.
```{r}
meiosis_output<-engage_in_meiosis(exampleData)
# head(meiosis_output) # as you can see: some of these have an impossible number of alleles (0.5 copies). The next function will fix that by generating gametes that have either 0 or 1 copies of the allele like in real meiosis.

compile_gametes_output<-compile_gametes(meiosis_output)
head(compile_gametes_output)
```

Let's look specifically at the genotypes of sperm, and then eggs.
```{r}
# now we can separate them into sperm and eggs
sperm_and_eggs(x = compile_gametes_output,sex = 'sex')
head(sperm)
head(eggs)
```
Now we can actually look at the output of fertilization. As you can see in the output of this functioin, we have genotypes, and we recorded who the mom and dad were. We also have a term called `momdad` that essentially saves that crossing (ie it pastes together the mom and the dad).
```{r}
fertilize_output<-fertilize(malegametes = sperm,
          femalegametes = eggs)

head(fertilize_output) 
```

Next, we have two ways of summarizing the outcome of the crossing. The first records the distribution of each genotype per momdad (ie per cross) and per gene.
```{r}
summarize_fertilization_output<-summarize_fertilization(fertilize_output)
head(summarize_fertilization_output)
```

The next way to summarize the data is probably more useful, and that's to generate the probability that each potential pup is actually born. Included in this summary are the probabilities a pup of that genotype is not born given a litter size from 1-10 pups. For example, "notthatgenotypeonepup" column gives the probability that, if 1 pup is born, what is the probability that it is not that genotype? 
```{r}
summarize_potential_pup_output<-summarize_potential_pups(fertilize_output)
head(summarize_potential_pup_output)
```

Let's say you have a specific goal of a genotype you want to create through breeding. Let's say we want homozygous positive mouse at 4 loci: AABBCCDD. Code this by making a vector of your desired genotype (let's call it your `desiredvec`). This desiredvec of genotypes needs to be the same length as the number of gene columns you have in your fertilize_output (which will be the same as what you supplied in the beginning to this walkthrough). Use the following words to make your desiredvec: homopos, het, homoneg. 

UPDATE YOUR DESIRED VECTOR BELOW APPROPRIATELY.
```{r}
desiredvec<-c('homopos', 'homopos', 'homopos', 'homopos')  # UPDATE THIS desiredvec APPROPRIATELY FOR YOUR NEEDS
```

First let's see if we can put all the alleles we want into one mouse in one cross (not necessarily getting 2 alleles (AA) if that's what you want, but getting at least one in as a heterozygote (Aa)). 
```{r}
can_we_get_all_the_alleles_from_one_cross(x = fertilize_output,desiredvector = desiredvec)
```

Next we can score each pup. The maximum total score for each pup is 100*(the number of loci or gene columns). Getting the genotype you want is worth 100, and getting a het if you want a homozygous is worth 50. Normalized points are the points the pup earned divided by the total possible points it could have achieved, all times 100.
```{r}
pointsperpupoutput<-points_per_pup(x = summarize_potential_pup_output)
head(pointsperpupoutput)
```

Finally, we can then ask the package for advice on which pairs to breed to get the highest possible normalized score.
```{r}
whichpairstobreed<-which_pairs_should_i_breed(x=pointsperpupoutput,desiredvector = desiredvec)
```
# Visualizations
## You may want to actually visualize this data, which you can do in some ways we recommend below.

First we wrangle the data a bit by melting it
```{r}
meltsummarizefertoutput<-melt(summarize_fertilization_output,
                              id.vars = c('gene','momdad'),
                              measure.vars = c('freqhomoneg','freqhet','freqhomopos'),
                              variable.name = 'genotype',
                              value.name = 'frequency'
)
meltsummarizefertoutput$percent<-(meltsummarizefertoutput$frequency)*100 # we calculate percent from the frequency
uniquemeltgenes<-unique(meltsummarizefertoutput$gene) # we look at the unique genes you gave earlier for plotting

plotlist<-list() # we make a new list
```

This chunk of code will plot separate plots for each gene, and show you the distribution of each genotype (homoneg, het, or homopos) for each crossing.
```{r}
for (w in uniquemeltgenes) {
  subsetmelt<-subset(meltsummarizefertoutput,meltsummarizefertoutput$gene==w)
  print(subsetmelt)
  plot_to_plot<-ggplot(subsetmelt)+geom_col(aes(y=percent,x=genotype,fill=genotype))+
    facet_wrap(~momdad)+
    scale_fill_viridis(labels=c("freqhomoneg" = "HomozygousNeg", "freqhet" = "Heterozygous",
                                "freqhomopos" = "HomozygousPos"),discrete = T,end=.8)+
    ggtitle(w)+
    theme(axis.text.x = element_text(angle=90),
          plot.title = element_text(hjust = 0.5))

  plotlist[[length(plotlist) + 1]] <- plot_to_plot

}
```

Depending on your number of crosses, this plot may be huge, and it may make more sense to look at each element of the list individually as shown in the first line below (change 1 to 2 etc to go through all the plots.).
```{r,fig.width=9,fig.height=60}
# plotlist[1] # we can peek at the first gene by calling the first element of the list
ggarrange(plotlist = plotlist,ncol=1) 
```

This next plot will do the same thing as above plot, but instead of making separate plots per gene, you can make separate plots per cross, which may be easier to read.
```{r,fig.width=9,fig.height=60}
rm(plotlist) # we remove this from your environment to make sure the list is empty before putting more things into it
plotlist<-list() # we make a new list
uniquemepairings<-unique(meltsummarizefertoutput$momdad) # pull the unique pairings we have
for (w in uniquemepairings) {
  subsetmelt<-subset(meltsummarizefertoutput,meltsummarizefertoutput$momdad==w)
  print(subsetmelt)
  plot_to_plot<-ggplot(subsetmelt)+geom_col(aes(y=percent,x=genotype,fill=genotype))+
    facet_wrap(~gene)+
    scale_x_discrete(labels=c("freqhomoneg" = "HomozygousNeg", "freqhet" = "Heterozygous",
                              "freqhomopos" = "HomozygousPos"))+
    scale_fill_viridis(discrete = T,end=.8)+
    ggtitle(w)+
    theme(axis.text.x = element_text(angle=90),
          plot.title = element_text(hjust = 0.5))

  plotlist[[length(plotlist) + 1]] <- plot_to_plot

}
ggarrange(plotlist = plotlist,ncol=1)
```

```{r}
# sessionInfo()
```

```{r}
### FIN ###
```

